File: src/DNA_parser.py
from blist import blist

#dna_type is something that supports following operations:
#    empty initialization
#    initialization from string
#    __len__()
#    [index]
#    [begin:end] -- produces dna_type as well
#    append(char)
#    extend(dna_type)
#  

dna_type = blist

class DNA_parser(object):
    def __init__(self, dna, freqs):
        assert isinstance(dna, dna_type)
        self.dna = dna
        self.dna_len = len(self.dna)
        self.index = 0  # needed outside the parser
        self.iter = iter(self.dna)
        self.saved_codon = None
        self.pattern_freqs, self.template_freqs, self.codon_len_freqs = freqs
     
    def read_base(self):
        '''return base or empty string in case of EOF'''
        assert self.saved_codon is None #because of the structure of the parser
        if self.index == self.dna_len:
            return ''
        self.index += 1
        return self.iter.next()
    
    def read_codon(self):
        '''return "", C, F, P, IC, IF, IP, IIC, IIF, IIP or III'''
        # it turned out to be useful for parsing pattern, template and consts
        if self.saved_codon is not None:
            result = self.saved_codon
            self.saved_codon = None
            return result
        
        dna = self.dna
        bases_left = self.dna_len - self.index
        max_len = bases_left if bases_left < 3 else 3 # faster than min
            
        codon = ''    
        codon_len = 0
        while codon_len < max_len:
            base = self.iter.next()
            codon_len += 1
            codon += base
            if base != 'I':
                break
               
        self.index += codon_len
        
        self.codon_len_freqs[codon_len] += 1
        return codon
    
    def unread_codon(self, codon):
        assert self.saved_codon is None
        self.saved_codon = codon

File: src/cpp_execute/DNAParser.cpp
#include <assert.h>

#include "DNAParser.h"

DNAParser::DNAParser(dna_type pdna) : pdna(pdna), index(0), iter(Iterator(pdna))
{
	dna_len = pdna->length();
	saved_codon = "";
}

t_base DNAParser::read_base()
{
	assert(saved_codon == "");
	if (!iter.valid)
		return 0;
	t_base result = iter.current();
	iter.advance();
	index++;
	return result;
}

t_codon DNAParser::read_codon()
{
	if (saved_codon != "")
	{
		t_codon result = saved_codon;
		saved_codon = "";
		return result;
	}

	int bases_left = dna_len - index;
	int max_len = bases_left < 3 ? bases_left : 3;

	t_codon codon = "";
	int codon_len = 0;
	while (codon_len < max_len)
	{
		t_base base = iter.current();
		iter.advance();
		codon_len += 1;
		codon += base;
		if (base != 'I')
			break;
	}
	index += codon_len;
	return codon;
}

void DNAParser::unread_codon(t_codon codon)
{
	assert(saved_codon == "");
	saved_codon = codon;
}

void DNAParser::jump(int offset)
{
	assert(offset >= 0);
	assert(offset < dna_len);
	saved_codon = "";
	index = offset;
	iter = Iterator(pdna);
	iter.advance(offset);
}

int DNAParser::nat()
{
	int result = 0;
	int power = 1;
	while (true)
	{
		t_base a = read_base();
		if ((a == 0) || (a == 'P'))
			break;
		if (a == 'C')
			result += power;
		power *= 2;
	}
	return result;
}

std::string DNAParser::consts()
{
	std::string result;
	while (true)
	{
		t_codon a = read_codon();
		if (a == "C")
			result += "I";
		else if (a == "F")
			result += "C";
		else if (a == "P")
			result += "F";
		else if (a == "IC")
			result += "P";
		else
		{
			unread_codon(a);
			break;				
		}
	}
	return result;
}
